File: src/algorithms/arrays/max-consecutive-ones/sources/MaxConsecutiveOnes.java
// Max Consecutive Ones III — O(n) variable sliding window with at most k zero-flips
public class MaxConsecutiveOnes {
    public static int[] maxConsecutiveOnes(int[] inputArray, int maxFlips) {
        if (inputArray.length == 0) { // @step:initialize
            return new int[]{0, 0}; // @step:initialize
        }

        int leftPointer = 0; // @step:initialize
        int zeroCount = 0;
        int maxLength = 0;
        int bestStartIndex = 0;

        // Expand the right boundary of the window
        for (int rightPointer = 0; rightPointer < inputArray.length; rightPointer++) {
            if (inputArray[rightPointer] == 0) {
                zeroCount++; // @step:expand-window
            }

            // Shrink from left when zero count exceeds the allowed flips
            while (zeroCount > maxFlips) { // @step:compare
                if (inputArray[leftPointer] == 0) {
                    zeroCount--; // @step:shrink-window
                }
                leftPointer++; // @step:shrink-window
            }

            int windowLength = rightPointer - leftPointer + 1; // @step:compare
            if (windowLength > maxLength) { // @step:compare
                maxLength = windowLength; // @step:compare
                bestStartIndex = leftPointer; // @step:compare
            }
        }

        return new int[]{maxLength, bestStartIndex}; // @step:complete
    }
}

File: src/algorithms/arrays/max-consecutive-ones/sources/max-consecutive-ones.py
# Max Consecutive Ones III — O(n) variable sliding window with at most k zero-flips
def max_consecutive_ones(input_array: list[int], max_flips: int) -> dict:
    if len(input_array) == 0:  # @step:initialize
        return {"max_length": 0, "start_index": 0}  # @step:initialize

    left_pointer = 0  # @step:initialize
    zero_count = 0
    max_length = 0
    best_start_index = 0

    # Expand the right boundary of the window
    for right_pointer in range(len(input_array)):
        if input_array[right_pointer] == 0:
            zero_count += 1  # @step:expand-window

        # Shrink from left when zero count exceeds the allowed flips
        while zero_count > max_flips:  # @step:compare
            if input_array[left_pointer] == 0:
                zero_count -= 1  # @step:shrink-window
            left_pointer += 1  # @step:shrink-window

        window_length = right_pointer - left_pointer + 1  # @step:compare
        if window_length > max_length:  # @step:compare
            max_length = window_length  # @step:compare
            best_start_index = left_pointer  # @step:compare

    return {"max_length": max_length, "start_index": best_start_index}  # @step:complete
